### PR TITLE
Use raw tracebacks to avoid functions with padding

### PR DESCRIPTION
**Pull request by marceloamaral**
After the discussion is the issue #1530, the option to avoid structure with padding is to use raw tracepoints
This PR is fixing that.

### CODE REVIEW

**Comment by marceloamaral** [comment]
Dave has already addressed this in PR #1542
I am closing it

**Comment by marceloamaral** [comment]
Reopening the PR since it is getting the task prev-state differently from the PR #1542
@rootfs can you test it?

**Comment by dave-tucker** [comment]
> Reopening the PR since it is getting the task prev-state differently from the PR #1542 @rootfs can you test it?

If you check the kernel code you can see how the args in the tracepoint progs are populated:
See: https://elixir.bootlin.com/linux/latest/source/include/trace/events/sched.h#L245

I don't think we should be accessing `__state` directly in `task_struct`.
[This function](https://elixir.bootlin.com/linux/latest/source/include/linux/sched.h#L1573) is used which calculates the state based on the value of `prev_state` and `prev_task->exit_state`. We might need to reimplement this in our code.

I can have a play around with bpftrace and compare the values we get from the `tracepoint` vs. `tp_btf` to see if they are consistent, or whether we need to do additional processing.

**Comment by rootfs** [comment]
here is a sample bpf sche_switch tracepoint usage
https://elixir.bootlin.com/linux/latest/source/tools/bpf/runqslower/runqslower.bpf.c

**Comment by dave-tucker** [comment]
I did some research and it turns out that the `prev_state` argument isn't part of the context of this probe on Kernel 5.12.
See: #1550 for the fix for that - which I've left as draft because 👇 

We should clear up whether or not we should be checking the task state at all - let's discuss that in #1546.

Depending on that outcome we'll know whether or not to proceed with #1550

**Comment by marceloamaral** [comment]
@rootfs @dave-tucker
Using the `state` of the `task_struct` follows the latest approach in the libbpf examples for collecting CPU time. This implementation can be seen in the cpudist.bpf.c example from the iovisor/bcc repository: [cpudist.bpf.c](https://github.com/iovisor/bcc/blob/6d3d8a2aca2772dbd91644462852206387e296f2/libbpf-tools/cpudist.bpf.c).

There have been many improvements and advances in CPU time collection in eBPF examples, and this method is the most reliable one.

**Review by dave-tucker — CHANGES_REQUESTED** [review]

File: bpfassets/libbpf/src/kepler.bpf.c [hunk 1]
```diff
@@ -194,28 +184,25 @@ static inline void collect_metrics_and_reset_counters(
 	buf->process_run_time = get_on_cpu_elapsed_time_us(prev_pid, curr_ts);
 }
 
-struct task_struct {
-	int pid;
-} __attribute__((preserve_access_index));
-
-SEC("tp_btf/sched_switch")
-int kepler_sched_switch_trace(u64 *ctx)
+SEC("raw_tp/sched_switch")
```
[the remark this answers is not in the document]
Comment (marceloamaral) [reply bpfassets/libbpf/src/kepler.bpf.c:187]: Switching to raw tracepoints can potentially reduce the overhead associated with BTF processing, but I am not sure if we really need that.

**Pull request closed without merging**